src/index.ts

```diff
@@ -1,61 +1,67 @@
-import { getTargetOrigin, TestBoxConfig } from "./config";
-import { info } from "./utils/logging";
-import {
-  isValidIncomingTestBoxMessage,
-  sendMessageToTestBox,
-} from "./messaging";
-import { routeMessage } from "./router";
+import type { TestBoxConfig, LoginHandler } from "./config";
+import { sendMessageToTestBox } from "./messaging";
 import { INITIALIZE_REQUEST } from "./messaging/outgoing";
 import { LoginEvent, NavigateEvent } from "./messaging/incoming";
+import { messageEventCallback } from "./message-event";
+import { routeMessage } from "./router";
 
 let tbxStarted = false;
+let isLoginHandlerRegistered = false;
+
+export let navigateHandler: (data: NavigateEvent) => void = (data) => {
+  window.location.href = data.url;
+};
+export let loginHandler: (data: LoginEvent) => Promise<string | boolean> =
+  undefined;
+function eventCallbackMiddleware(ev: MessageEvent<any>) {
+  messageEventCallback(ev, { navigateHandler, loginHandler });
+}
+
+export async function registerLoginHandler(newLoginHandler: LoginHandler) {
+  if (isLoginHandlerRegistered) {
+    throw new Error("LoginHandler already registered!");
+  }
 
-export let navigateHandler: (data: NavigateEvent) => void;
-export let loginHandler: (
-  data: LoginEvent
-) => Promise<string | boolean> = async () => false;
+  if (!tbxStarted) {
+    throw new Error(
+      "StartTestbox function must be called before registering login handler!"
+    );
+  }
+
+  window.removeEventListener("message", eventCallbackMiddleware);
+
+  loginHandler = newLoginHandler;
+
+  if (window?.__tbxLoginEvent) {
+    routeMessage(window?.__tbxLoginEvent, {
+      navigate: navigateHandler,
+      login: loginHandler,
+    });
+    window.__tbxLoginEvent = null;
+  }
+
+  window.addEventListener("message", eventCallbackMiddleware);
+
+  isLoginHandlerRegistered = true;
+}
 
 export function startTestBox(config?: TestBoxConfig) {
   if (tbxStarted) {
     return;
   }
 
   window.__tbxConfig = config;
+  window.__tbxLoginEvent = null;
 
   if (window.__tbxConfig.navigateHandler) {
     navigateHandler = window.__tbxConfig.navigateHandler;
-  } else {
-    navigateHandler = (data) => {
-      window.location.href = data.url;
-    };
   }
```

Comment:
We currently have no tests and no requirements reg the navigation handler, so better to avoid any changes on that front.

Any reason that this was changed?

Reply:
I moved this function definition to the variable on the top of the file